### DISCUSSION TITLE
Efficient way of combining xarray datasets (explode + combine_by_coords) with variables only sharing subset of coordinates

### DISCUSSION THREAD

**Discussion by adrien-perello** [pydata/xarray#6029] [Q&A]
(message edited for clarity)

## Context

I'm running a simulation that generates multiple (networkx) graphs. (It's actually a [mesa](https://mesa.readthedocs.io/en/stable/) agent-based simulation inspired from the [Virus on a Network](https://github.com/projectmesa/mesa/tree/main/examples/virus_on_network) example) and due to randomness, I run each set of parameters multiple times.

## Minimal example

The following example hopefully gives an idea of what I'm trying to achieve

```python
import numpy as np
import pandas as pd
import networkx as nx

def avg_degree(G):
    return 2 * G.number_of_edges() / G.number_of_nodes()

def degree_distribution(G):
    return np.array(nx.degree_histogram(G))

def network_metrics(G):
    return {
        "avg degree": avg_degree(G),
        "degree distribution": degree_distribution(G),
    }

def generate_data(step, run, n, p):
    G = nx.erdos_renyi_graph(n, p)
    dct = {
        'network_type': 'random',
        'run': run,
        'nb_agents': n,
        'probability': p,
        'infected': np.random.randint(0,100, step),
    }
    dct.update(network_metrics(G))
    return dct


def main(nb_steps):
    lst = []
    for run in range(4):
        for nb_nodes in [10, 20, 50]:
            for probability in [0.1, 0.5, 0.8]:
                dct = generate_data(nb_steps, run, nb_nodes, probability)
                lst.append(dct)

    result = pd.DataFrame(lst)
    indexes = ['network_type', 'run', 'nb_agents', 'probability']
    result.set_index(indexes, inplace=True, drop=True)
    return result
```

**Which gives**

![Screenshot from 2021-11-26 15-51-02](https://user-images.githubusercontent.com/76055749/143598852-99f3cf82-52da-45c3-af15-c4996ae7fd8e.png)

## Goal

1) Explode **`infected`** and **`degree distribution`** (!! note: `infected` has a fixed length (= `nb_steps`) but not `degree distribution` (whose length varies)).
2) Combine everything in a single dataset 

## Current solution

I use the following helper function to explode the different columns 

```python
def explode(df, mapping):
    new_df = df.reset_index()
    for col, idx in mapping.items():
        new_df.index.rename('_id', inplace=True)
        new_df = new_df.explode(col)
        new_df.insert(1, idx, new_df.groupby('_id').cumcount())
        new_df.reset_index(drop=True, inplace=True)

    idx = list(df.index.names)
    nested_idx = list(mapping.values())
    return new_df.set_index(idx + nested_idx)


def helper(df, mapping):
    sol = []
    for k,v in mapping.items():
        sol.append(explode(df[k], {k:v}).to_xarray())

    leftover_columns = list(df.columns.difference(mapping.keys()))
    sol.append(df[leftover_columns].to_xarray())
    return sol
```

which I use as follow:

```python
mapping = {'infected': 'step', 'degree distribution': "nb_nodes_with_degree"}
lst = helper(result, mapping)
xr.combine_by_coords(lst)
```

## Result

![Screenshot from 2021-11-26 16-16-40](https://user-images.githubusercontent.com/76055749/143601583-1a0b2f9e-9cfe-4837-b13b-b6705a9d547a.png)

![Screenshot from 2021-11-26 16-11-19](https://user-images.githubusercontent.com/76055749/143600921-2c139d78-d346-425c-a055-e19afb9ade29.png)

## Limits

It works, but it is quite slow and really inelegant.
--> What would be a better way of handling it?

**Comment by max-sixty** [comment]
Coordinates can be any number of dimensions!

Can you give more detail if that doesn't answer your question, ref https://github.com/pydata/xarray/discussions/5404

**Reply by adrien-perello** [reply]
(message deleted for clarity --> see original message)

**Reply by max-sixty** [reply]
Hi Adrien — I'd really encourage you to reduce your example much more to a targeted xarray question, use paste-able code — all the items in the linked issue.

**Reply by max-sixty** [reply]
Thanks for the edit! That's much clearer!

**Reply by max-sixty** [reply]
I'm going to respond quickly rather than completely, but we can go back & forth:

Working on data in lists and then using something like `df.explode` is very slow in python, relative to having the data in array form.

One of the advantages of xarray is being able to have an arbitrary number of dimensions — so `infected` shouldn't need to be stored as a python list IIUC, since it's always the same size — instead that can be stored as a variable with an additional dimension. `degree_distribution` may be more difficult, since it doesn't have a consistent size.

The profile you posted is excellent. The only surprising thing to me is how long `to_xarray` takes. If you can turn that specific operation into a minimal example (i.e. without `nx`, and ensuring none of the values are python lists) and you still see poor performance, that would be a bug.

**Comment by adrien-perello** [comment]
using `cProfile` and `gprof2dot` gives an idea of the bottlenecks. And it actually comes from the `explode()` together with `to_xarray()`

![output](https://user-images.githubusercontent.com/76055749/143611568-a038efd4-5819-4d48-b31b-d093a98bacb3.png)
